docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,414 @@
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+      oe_get_evidence(&format, 0, claimsY, countof(claimsY), NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   oe_claim_t* claims;
+   size_t claims_length;
+   oe_get_verifier_claims(&format, &claims, &claims_length);
+
+   oe_get_evidence(&format, 0, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+```
+
+4. SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. The local/remote attestation concept is SGX specific. To prevent confusion at evidence generation, the OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION and OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION flags were deprecated.
+
+```C
+   // Old
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, claims, claims_length, target, target_size, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+   oe_get_evidence(&format_remote_only, OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL); // Error
+
+   // New
+   oe_get_evidence(&format_sgx_local, 0, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_sgx_remote, 0, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_local_only, 0, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+   oe_get_evidence(&format_remote_only, 0, claims, claims_length, NULL, 0, &evidence, &evidence_size, NULL, NULL);
+```
+
+User Experience
+---------------
+The user experience is improved by the changes. These are a couple of possible attestation scenarios:
+
+Scenario 1 - Attester selects from a list of evidence formats that can be verified:
+- Verifier application
+    1. Upon initialization, queries available verifiers to the application (oe_sgx_get_verifiers) and registers them (oe_register_verifier).
+    2. Builds the format IDs of the registered verifiers sorted by preference.
+    3. Sends the format IDs to the attester application. 
+- Attester application
+    1. Upon initialization, queries available attesters to the application (oe_sgx_get_attesters) and registers them (oe_register_attester).
+    2. Receives format IDs from the verifier application.
+    3. Selects a format ID from the received format IDs (oe_select_attester_format).
+    4. Sends the format ID to the verifier application.
+- Verifier application
+    1. Gets verifier claims for the received format ID (oe_get_verifier_claims).
+    2. Sends the claims to the attester application.
```

Comment:
I think in this step, what the verifier sends over to the attester should be optional parameter, and not a set of claims.
Use SGX local attestation as an example, what the verifier enclave needs to provide to the attester enclave is a structure with the verifier's SGX target info. This structure should be used by the attester as the optional parameter in its call to `oe_get_evidence()`.

To help clarify what goes to  custom claims and what goes to optional parameter, it would be good to articulate the semantics of these two fields, both in this design doc and in the OE SDK user's guide.

Here is my take on their semantics, and why the SGX target info is an optional parameter:

- Custom claims have well-defined semantics:
  - Can be any name-value pair
  - Only meaningful to ISV attester / verifier application
    - Typically, custom claims are part of the attester instance identity or properties. They are created and maintained by the attester and not received from a verifier
    - Examples include attester public key, or a nonce generated by attester
    - These claims are passed to verifiers with integrity protection (as part of the evidence), so verifiers can use the claims to learn the identity and properties for the attester and for secure communication with the attester
  - They are always TEE agnostic
  - Signed and included in evidence as clear text by an attester plugin, and verified and reproduced by a verifier plugin, to ensure integrity protection in transmission from attester application to verifier application
    - Secret shall never be included as a custom claim. Instead a hash of a secret can be included as a claim
  - Custom claims are NOT inspected or interpreted by OE SDK / plugins
- Optional parameters semantics are the opposite of that of custom claims:
  - They are TEE and evidence format specific
  - They are interpreted and processed by OE SDK attester and verifier plugins
    - Typically, optional parameter is provided by a verifier to an attester, so the attester can generate the correct evidence that can be successfully verified by the verifier
  - They are not interpreted nor processed by ISV attester / verifier application
- SGX target info is perfect fit for optional parameter
  - For SGX local attestation, the SGX target info is generated by a verifier enclave and passed to an attester enclave
  - With the received target info, the attester can generate an SGX report targeted for the verifier enclave
  - This SGX report can only be verified by the verifier enclave that provided by target info. Any other enclaves, either running on the local or a remote platform, won’t be able to verify this SGX report.

Reply:
I checked RATS ([https://www.ietf.org/id/draft-birkholz-rats-architecture-03.txt](https://www.ietf.org/id/draft-birkholz-rats-architecture-03.txt)) and the definitions there align with your statement. Claims should not flow from verifier to attester. Fixed.

---
In reply to: [391267166](https://github.com/openenclave/openenclave/pull/2621#discussion_r391267166) [](ancestors = 391267166)

Reply:
FYI, the draft-Birkholz document is not the correct document (that one is largely obsolete and was never adopted by the WG).  The correct document is https://tools.ietf.org/html/draft-ietf-rats-architecture